### PR TITLE
Pjax jQuery templates

### PR DESCRIPTION
**Pull request by notsambeck**
Hey all:

Here is the site built around [pjax](https://github.com/defunkt/jquery-pjax). Only extremely minor content changes - it's 95% just restructuring, most of that to main.js and in headers/footers on each page.

What I did:

1. Implemented templates and client-side only pjax. Headers and footers are no longer duplicated, and load only once per site visit. JS/CSS also is requested/loaded only once. Falls back to standard HTML if not supported.
2. Re-indented all html for consistency and readability. Unfortunately, this makes the github diff tool pretty much worthless - sorry I should have thought about this beforehand.
3. Moved to the newest version of jQuery & made needed changes; this had already happened on some pages.
4. Minor css things to improve the mobile site (mostly removing empty space on index.html).
5. Cleaned up .js, 90% now lives in main.js
6. Deleted three redundant pages rather than deal with them: 
- repositories.html 
- about.html
- template.html

The site works well locally on my Linux desktop in Firefox and Chrome, so that is good!

**Possible issues** that need consideration before deploying to Google hosting:
- pjax events are logged to the console for debugging - we probably should turn that off? all in main.js
- There's no way for me to know if Google Analytics is actually working
- Navigation active links are based on URL, so we will need to confirm they still work on GAE
- Before deployment we could replace jquery-3.2.1.js, bootstrap.js/.css with minimized versions?

I have taken the liberty of putting this up as a github.io page for some mobile testing: [github pages site](https://notsambeck.github.io/).
There are a few issues with the github.io-hosted site that I think will go away when hosted on GAE - it's code that I haven't changed. **Alternatively - it is possible to redirect from www.openworm.org to a GitHub.io site which would allow for easy testing (by anyone) on an identical environment to the actual deployment. GitHub pages seems like a better long-term solution to me. **

These issues are trivial to fix - i resolved them on the notsambeck.github.io branch but we will need to change references from http to https on the main branch before deploying.
Known issues that will be solved this way:
- Error message - 'blocked loading mixed active content': GitHub widget & videos on get-started page do not load
- Home link on nav bar does not show as active on browser back

Let me know if you guys have questions!

### CODE REVIEW

**Comment by notsambeck** [comment]
See issue #93

**Comment by notsambeck** [comment]
@tarelli thanks for the feedback! Those are all issues that I considered, here is my thinking on each of them. If the consensus is that we should change any of these decisions, let's do it! I won't be offended.

**about.html**
If someone wants to know generally about the project, where should they go?
- home
- get started.html
- documentation -> introduction
- (about.html)
In my mind, this is too many options - it makes it hard for the user to know what to do. It also makes it hard for us to keep messaging consistent across the site. I think that the messaging on get_started is the strongest version of what we have. get_started.html could be renamed about.html, if anything.

**repositories.html**
It seems that there are no links to this page? I'm only seeing links in dead kickstarter pages
Also, this information seems to be duplicated (manually) in several places:
- documentation -> repos
- github.com
- (repositories.html)
My thought (again) is that it's best to minimize that repetition, put things in the single best place, and cut down number of pages to maintain. 

**pjax / template strategy in general**
tldr: With pjax, each page is self-contained and needs to have everything an html page needs. The advantage is, navigation is simple and maintainable.

I spent a lot of time looking into this. I see two poles of how a site can be designed:

1. pure html; each page is a standalone object. 
The advantage of an old-fashioned html site is that navigation is between discrete pages/urls, all of which are accessible & searchable from any location
The disadvantage is code repetition on each page and (possibly) performance.

2. single-page-app; there is literally only one page, and content is loaded into it on demand.
The advantage of a single-page app is it eliminates repeated code.
The disadvantages are:
- If browser doesn't support this correctly, what happens?
- You have to re-invent navigation. For example, if you want to start your visit on people.html, what does that mean? Probably there is a page people.html that redirects to index.html#people. But what happens if you then hit back in your browser?
- For a site that is maintained by volunteers, this reinvented navigation code seems likely to become complicated and outweigh the benefits of minimizing repeated code.


I went with a solution in the approximate middle of these poles, trying to follow [ROCA](http://roca-style.org/) as a guide. Basic principles: keep navigation in the browser. My design definitely isn't perfect, but I think it solves all of the design problems reasonably well:
- Repeated header/footer code only loads once, so for the user it functions just like a single page app (fast/seamless)
- Each page is fully self-sufficient, so A. you can start visit on any page B. you can link to any page C. browser back/forward/reload work normally
- If browser doesn't support pjax, it falls back to standard html with no issues (and no alternate site to maintain).
- The downside of that self-sufficiency is, each page needs to have all those things (<html> tag, css, js loaders). 

It would be possible to load those things with {{}} or jQuery - but you still have to load the loader. It would make the system more fragile. You could argue the same about the navigation header and footer (that they should be repeated in full on each page), but I decided that haeder/footer are complex enough that it's worth the trade-off to have them not repeat. Also, because this implementation doesn't use any server-side logic, server the whole page and the browser pulls out the pjax (body) content - so by linking to headers/footers, that content is not served except on the initial request (i.e. the jQuery that loads them is only executed once).

Anyway that's what I was thinking! Let me know what you think.

**Comment by gidili** [comment]
@notsambeck @tarelli my opinion: 

First of all thanks for the effort @notsambeck 👍 

The repositories page is a side-project we were working on with @cheelee and that's why it's not linked anywhere, it's work in progress, a page to see all repositories and documentation etc. so it would be nice if it doesn't get wiped 😂 

For the about page I don't feel strongly about it staying or going but I think the contents that we have in there (what, why who) should be incorporated in some other page (maybe simply a new section in getting started) because it's pretty much the best summary of the project we have. 

The templates page is something we had in there as a template for a page to copy-paste when creating a new page, rather than delete it we'd need an updated one with the new stuff (header and footer injected etc).

About the templating engine, what you have done is great and a big improvement - it would be nice if we can basically avoid duplicating html stuff in general applying the same strategy you applied to the header / footer to actual html sections that are still repeated identical everywhere. It is currently not clear to me if with pjax it can be achieved or not (question)?

**Comment by notsambeck** [comment]
@gidili Aha! Thanks for info. 
cc: @tarelli 

- I will resurrect the repositories page - sounds pretty slick! 
- I will also look carefully at the about/get_started stuff and see if I can combine them effectively.

RE: pjax / code repetition
It would be possible to put lines into main.js to load css from there, similar with favicons, etc.:
$("head").append("<link rel='stylesheet' type='text/css' href='whatever.css')

However, I feel like that would be less clean than having explicit links to the css on each page. By delaying the .css load until after .js executes, the page might reformat on initial load which would be weird.

No matter what with pjax, each page will need to include 
1. html tags
2. linked script to run the .js

It would be possible to reduce the number of repeated lines, but not that there are repeated lines. So in my mind there's not a lot to gain by hiding those references elsewhere.

I do want to go through the .css (as @geekSiddharth and I were talking about the other day) and clean it up substantially, reduce number of files linked - but that's another topic.

**Comment by notsambeck** [comment]
Hey all: 

I did those little fixes I said I would do:
1. Brought back repositories.html (file is unchanged; currently does not work on github)
2. Added a section to index.html using highlights from copy in about.html; minor changes to spacing in that section. I think adding this little bit of content helps quite a bit.
3. Fixed http/https issues so everything loads on Github.io

I copied the /war/ folder and posted it to my github.io site again for preview purposes: [notsambeck.github.io](https://notsambeck.github.io/index.html) 

Let me know if you have questions or issues!
Sam

**Comment by tarelli** [comment]
@notsambeck Hey Sam, I checked the latest, nice improvements, may I suggest the following: put "OpenWorm is an open source project dedicated to creating the first virtual organism in a computer." under the "why" instead of the current text "OpenWorm has been developing a complete cell-level model of the C. elegans nematode since 2011." and restore the logo to be a bit bigger and centered.
Also there seems to be a problem on the margins around the news:
screen shot 2017-07-27 at 10 30 17

Regarding the templates what I had in mind is something along the lines of what's described here using mustache.js https://www.sitepoint.com/creating-html-templates-with-mustachejs/.

screen shot 2017-07-27 at 10 48 55

Basically only one template file with HTML block and css/js common to all pages. Then one file for every page which only has whatever content would be injected inside the body of the template (this is how template engines ususally work, django, ejs, etc.).The body could have the menu bar in common to all and define two placeholders: headerContent (the green box at the top) and mainContent. Every page would then specify only what to put inside headerContent and what to put inside mainContent.
The advantages of this type of structure is that there is only one place, so say whatever planned work we do in terms of reducing the number of CSS we have and whatnot would be done only in the main template rather than having to go and change it in every page. I know this would require a bit more work but I wanted to make sure at least I'm communicating well what I've in mind. Maybe we can discuss this on a quick call one of these days @notsambeck @gidili  @geekSiddharth and talk through pros and cons?  Thanks again for doing this btw :)

**Comment by notsambeck** [comment]
Hey y'all - 

Sorry I vanished for a minute there. I fixed the clear issues from @tarelli 's comment above, so I think this is ready go live if you want to do that. Works on github pages hosting.

Regarding mustache - @tarelli thanks for spelling out your thoughts on that clearly. That makes sense, but I don't think it's worth the complexity trade off for a static site. The examples for templating are always repeated data forms (lists of tweets, lists of people, etc.). I don't think this short set of web pages would benefit the same way. For the end user, it would be exactly the same experience.

Let me know if you have questions!
Sam
